### PR TITLE
Add support for cross-platform builds including linux/amd64 and linux/arm64/v8

### PR DESCRIPTION
**Pull request by jtbaker**
Another attempt at the earlier #27, now that there is better multi-platform Docker support from the ecosystem.

~This PR simplifies and consolidates all of the logic in build-and-publish.sh into the GH Action, and uses the Docker [build-push-action](https://github.com/docker/build-push-action)~ 

This PR modifies the build script to use `docker buildx` to simplify the process of doing multi-platform builds.

Decided to have another go at this now that ~there is a built-in GH action that can be used for multi-platform builds, and~ current versions of maven are available as cross platform base images on Docker Hub.

https://github.com/IsraelHikingMap/graphhopper-docker-image-push/issues/26#issuecomment-2468959109

I think this will be useful to have as more and more cloud providers are offering ARM64 based VMs

https://cloud.google.com/compute/docs/general-purpose-machines#c4a_series
https://aws.amazon.com/ec2/graviton/

### CODE REVIEW

**Comment by HarelM** [comment]
Can we keep most of the build logic in the sh file to allow local testing and building?

**Comment by jtbaker** [comment]
@HarelM rewired the build script. Testing locally on my M1 mac, it does seem to be working, although building for the non-native platform (arm64 in my case) does take quite a bit longer.

File: build.sh [hunk 1]
```diff
@@ -47,11 +47,19 @@ if [ "$1" ]; then
   (cd graphhopper; git checkout --detach "$1")
 fi
 
-echo "Building docker image ${imagename}"
-docker build . -t "${imagename}"
+# Create and use a new builder instance for multi-platform builds
```
Comment (HarelM) [inline build.sh:50]: Use echo instead of comments.

File: build.sh [hunk 2]
```diff
@@ -47,11 +47,19 @@ if [ "$1" ]; then
...
+docker buildx create --use --name graphhopperbuilder
 
-if [ "${push}" == "false" ]; then
-  echo "Use \"docker push ${imagename}\" to publish the image on Docker Hub"
+echo "Building docker image ${imagename} for linux/amd64 and linux/arm64/v8"
```
Comment (HarelM) [inline build.sh:53]: I would consider removing this echo and improve the below echos according to what they actually do now.

**Comment by HarelM** [comment]
Added a few minor comments. 
Did you check that this is working as expected?

File: build.sh (hunk 2 above)
Comment (jtbaker) [reply build.sh]: addressed

File: build.sh (hunk 1 above)
Comment (jtbaker) [reply build.sh]: addressed

**Review by HarelM — APPROVED** [review]

**Review by jtbaker — COMMENTED** [review]
> Added a few minor comments. 
> Did you check that this is working as expected? 

@HarelM I have run the ./build.sh script locally and it seems to work, although it is still running for the amd64 version after ~3800s, so can't confirm 100% yet:
```
=> [linux/amd64 build 4/4] RUN mvn clean install                                                                        3829.6s
```

I also can't test providing the `--push` functionality to the repo since I don't have the permissions to the registry.

**Comment by HarelM** [comment]
I've manually triggered a build here:
https://github.com/IsraelHikingMap/graphhopper-docker-image-push/actions/runs/12038946704/job/33565737670

Not sure if it's stuck or just takes longer to complete.
Please monitor if this is working, otherwise I'll need to revert this PR...

**Comment by jtbaker** [comment]
> I've manually triggered a build here: https://github.com/IsraelHikingMap/graphhopper-docker-image-push/actions/runs/12038946704/job/33565737670
> 
> Not sure if it's stuck or just takes longer to complete. Please monitor if this is working, otherwise I'll need to revert this PR...

I'll check back this evening. If this doesn't work, can revert and look into having a separate step or possibly entirely different GH action run using a native runner for each platform, since it looks like Actions have Arm64 runners available now: https://github.blog/changelog/2024-09-03-github-actions-arm64-linux-and-windows-runners-are-now-generally-available/

**Comment by HarelM** [comment]
Please open a PR to either fix or revert this change. You can also fork this repo and try out to see what passes and what doesn't.
I don't think the problem is with the shell script, it's pretty basic...

**Comment by jtbaker** [comment]
@HarelM changes reverted in #39 

I don't think the problem is with doing the shell script per se, it probably is more to do with running a heavyweight Java app through the virtualization layer in docker buildx on a non-native platform (building the arm64 image on a amd64 machine and vice versa).

I will do some more banging on it in my fork at some point and see if I can get it stable and working.

**Comment by HarelM** [comment]
Sure thing, let me know what you find out.

**Pull request merged**